docs/proposals/20191017-control-plane.md

```diff
@@ -0,0 +1,494 @@
+
+}
+
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=controlplanes,shortName=cp,scope=Namespaced,categories=cluster-api
+// +kubebuilder:storageversion
+// +kubebuilder:subresource:status
+// +kubebuilder:subresource:scale:specpath=.spec.replicas,statuspath=.status.replicas,selectorpath=.status.selector
+
+// ControlPlane is the Schema for the ControlPlane API
+type ControlPlane struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec   ControlPlaneSpec   `json:"spec,omitempty"`
+    Status ControlPlaneStatus `json:"status,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// ControlPlaneList contains a list of ControlPlane
+type ControlPlaneList struct {
+    metav1.TypeMeta `json:",inline"`
+    metav1.ListMeta `json:"metadata,omitempty"`
+    Items           []ControlPlane `json:"items"`
+}
+
+func init() {
+    SchemeBuilder.Register(&ControlPlane{}, &ControlPlaneList{})
+}
+
+```
+
+With the following validations:
+
+- If the referenced InitTemplate does not define external etcd (webhook):
+  - ControlPlane.Spec.Replicas in [1, 3, 5, 7]
+- ControlPlane.Spec.Version != "" (openapi)
+
+And the following defaulting:
+
+- ControlPlaneSpec.Replicas: 1 (openapi or webhook)
+
+#### Modifications required to existing API Types
+
+- Add Cluster.Spec.ControlPlaneRef defined as:
+
+```go
+    // ControlPlaneRef is an optional reference to a provider-specific resource that holds
+    // the details for provisioning the Control Plane for a Cluster
+    // +optional
+    ControlPlaneRef *corev1.ObjectReference `json:"controlPlaneRef,omitempty"`
+```
+
+- Add Cluster.Status.ControlPlaneReady defined as:
+
+```go
+    // ControlPlaneReady defines if the control plane is ready
+    // +optional
+    ControlPlaneReady bool `json:"controlPlaneReady"`
+```
+
+#### Behavioral Changes from v1alpha2
+
+- If Cluster.Spec.ControlPlaneRef is set:
+  - [Status.ControlPlaneInitialized](https://github.com/kubernetes-sigs/cluster-api/issues/1243) is set based on the value of Status.Initialized for the referenced resource
+  - Status.ControlPlaneReady is set based on the value of Status.Ready for the referenced resource, this field is intended to eventually replace Status.ControlPlaneInitialized as a field that will be kept up to date instead of set only once.
+- Current behavior will be preserved if Cluster.Spec.ControlPlaneRef is not set
+
+#### Behaviors
+
+##### Create
+
+- A created ControlPlane should bootstrap a control plane with a given number of replicas
+- If an error occurs, errorStatus and errorReason will be defined on the ControlPlane object
+- Cannot create an even number of stacked control planes or anything other than 1, 3, 5, or 7
+- Can create an arbitrary number of control planes if etcd is external to the control plane
+- Creating a ControlPlane with > 1 replicas is equivalent to creating a ControlPlane with 1 replica followed by scaling the ControlPlane to 3 replicas
```

Comment:
I don't know, I think this is an important detail. CAPV relies on this today to know that the initial control plane node comes online and is reconciled with the core controllers before the bootstrap provider will create bootstrap data for additional nodes, control plane or otherwise.

I think bringing up an initial node by itself is an important behavior that should be maintained.

Reply:
The main purpose of this, in addition to calling out support for the use case @akutz mentions, is that this ties into the provided diagrams, which are simplified to show creation as just bringing up a single control plane machine, and the additional machines are brought up using the scale up workflow.

Reply:
This is fine to keep, I wanted to call out that this stands out because it dictates an internal detail, while the others are worded as simple requirements

Reply:
Perhaps it can be reworded to make it more clear that it is a required behavior?

Reply:
I'm going to defer this to a later iteration, currently this is mainly to simplify the proposal and related diagrams rather than a long term requirement on the controller.